File: week-06/cpp/01_multiple_inheritance.cpp
/**
 * Week 6 - Multiple Inheritance in C++
 */

#include <iostream>
#include <string>

using namespace std;

// Example 1: Simple Multiple Inheritance
class Flyable {
public:
    void fly() {
        cout << "Flying in the air" << endl;
    }
};

class Swimmable {
public:
    void swim() {
        cout << "Swimming in water" << endl;
    }
};

class Animal {
protected:
    string name;
public:
    Animal(string n) : name(n) {}
    
    void eat() {
        cout << name << " is eating" << endl;
    }
};

class Duck : public Animal, public Flyable, public Swimmable {
public:
    Duck(string n) : Animal(n) {}
    
    void quack() {
        cout << name << " says: Quack!" << endl;
    }
};


// Example 2: Diamond Problem with Virtual Inheritance
class Base {
public:
    int value;
    Base() : value(0) {
        cout << "Base constructor" << endl;
    }
    
    virtual void display() {
        cout << "Base value: " << value << endl;
    }
};

// Virtual inheritance to solve diamond problem
class Left : virtual public Base {
public:
    Left() {
        cout << "Left constructor" << endl;
    }
};

class Right : virtual public Base {
public:
    Right() {
        cout << "Right constructor" << endl;
    }
};

class Bottom : public Left, public Right {
public:
    Bottom() {
        cout << "Bottom constructor" << endl;
    }
};


int main() {
    cout << string(60, '=') << endl;
    cout << "Multiple Inheritance - Duck Example" << endl;
    cout << string(60, '=') << endl;
    
    Duck duck("Donald");
    duck.eat();   // From Animal
    duck.fly();   // From Flyable
    duck.swim();  // From Swimmable
    duck.quack(); // From Duck
    
    cout << "\n" << string(60, '=') << endl;
    cout << "Diamond Problem with Virtual Inheritance" << endl;
    cout << string(60, '=') << endl;
    
    Bottom b;
    b.value = 42;  // Only one copy of Base due to virtual inheritance
    b.display();
    
    cout << "\nWithout virtual inheritance, there would be ambiguity!" << endl;
    cout << "Virtual inheritance ensures only one Base subobject." << endl;
    
    return 0;
}

File: week-06/python/01_multiple_inheritance.py
"""
Week 6 - Multiple Inheritance in Python
"""

# Example 1: Simple Multiple Inheritance
class Flyable:
    """Mixin for flying capability"""
    
    def fly(self):
        print(f"{self.__class__.__name__} is flying")


class Swimmable:
    """Mixin for swimming capability"""
    
    def swim(self):
        print(f"{self.__class__.__name__} is swimming")


class Animal:
    """Base animal class"""
    
    def __init__(self, name):
        self.name = name
    
    def eat(self):
        print(f"{self.name} is eating")


class Duck(Animal, Flyable, Swimmable):
    """Duck inherits from Animal, Flyable, and Swimmable"""
    
    def __init__(self, name):
        super().__init__(name)
    
    def quack(self):
        print(f"{self.name} says: Quack!")


# Example 2: Diamond Problem and MRO
class A:
    def process(self):
        print("Process in A")


class B(A):
    def process(self):
        print("Process in B")
        super().process()


class C(A):
    def process(self):
        print("Process in C")
        super().process()


class D(B, C):
    """Diamond inheritance: D -> B -> A and D -> C -> A"""
    
    def process(self):
        print("Process in D")
        super().process()


if __name__ == "__main__":
    print("=" * 60)
    print("Multiple Inheritance - Duck Example")
    print("=" * 60)
    
    duck = Duck("Donald")
    duck.eat()  # From Animal
    duck.fly()  # From Flyable
    duck.swim()  # From Swimmable
    duck.quack()  # From Duck
    
    print("\n" + "=" * 60)
    print("Diamond Problem and MRO")
    print("=" * 60)
    
    d = D()
    d.process()
    
    print("\nMethod Resolution Order for class D:")
    print(D.__mro__)
    
    print("\nMRO in readable format:")
    for i, cls in enumerate(D.__mro__):
        print(f"{i}: {cls.__name__}")
